File: Alkitab/src/yuku/alkitab/base/util/QueryTokenizer.java
package yuku.alkitab.base.util;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class QueryTokenizer {
	public static final String TAG = QueryTokenizer.class.getSimpleName();
	
	static Pattern oneToken = Pattern.compile("(\\+?)((?:\".*?\"|\\S)+)"); //$NON-NLS-1$

	/**
	 * Convert a query string into tokens. Takes care of quotes and plus signs.
	 * Put quotes between words to make it one token, and optionally put plus sign before a token, to make it "whole phrase"
	 * 
	 * Examples:
	 *   a b => 'a', 'b'
	 *   "a b" c => 'a b', 'c'
	 *   a"bc"d => 'abcd' 
	 *   +"a bc" => '+a bc'
	 *   +a bc => '+a', 'bc'
	 *   +a+b => '+a+b'
	 *   
	 * @return tokens with plus still intact but no quotes.
	 */
	public static String[] tokenize(String query) {
		List<String> raw_tokens = new ArrayList<String>();

		Matcher matcher = QueryTokenizer.oneToken.matcher(toLowerCase(query));
		while (matcher.find()) {
			raw_tokens.add(matcher.group(1) + matcher.group(2));
		}
		
		//# process raw tokens
		List<String> processed = new ArrayList<String>(raw_tokens.size());
		for (int i = 0, len = raw_tokens.size(); i < len; i++) {
			String token = raw_tokens.get(i);
			if (token.length() > 0 && QueryTokenizer.tokenWithoutPlus(token).length() > 0) {
				processed.add(token.replace("\"", "")); // remove quotes //$NON-NLS-1$ //$NON-NLS-2$
			}
		}
		return processed.toArray(new String[processed.size()]);
	}

	/**
	 * Implementation of toLowerCase for Latin A-Z only.
	 * This is done because the Bible text is also lowercased using this method for searching
	 * instead of using the more sophisticated {@link String#toLowerCase()} method.
	 */
	static String toLowerCase(final String s) {
		final char[] newString = new char[s.length()];
		for (int i = 0, len = s.length(); i < len; i++) {
			final char c = s.charAt(i);
			if (c >= 'A' && c <= 'Z') {
				newString[i] = (char) (c | 0x20);
			} else {
				newString[i] = c;
			}
		}
		return new String(newString);
	}

	static boolean isPlussedToken(String token) {
		return (token.startsWith("+")); //$NON-NLS-1$
	}

	static String tokenWithoutPlus(String token) {
		int pos = 0;
		int len = token.length();
		while (true) {
			if (pos >= len) break;
			if (token.charAt(pos) == '+') {
				pos++;
			} else {
				break;
			}
		}
		if (pos == 0) return token;
		return token.substring(pos);
	}
	
	static boolean isMultiwordToken(String token) {
		int start = 0;
		if (isPlussedToken(token)) {
			start = 1;
		}
		for (int i = start, len = token.length(); i < len; i++) {
			char c = token.charAt(i);
			if (! (Character.isLetter(c) || ((c=='-' || c=='\'') && i > start && i < len-1))) {
				return true;
			}
		}
		return false;
	}
	
	static Pattern pattern_letters = Pattern.compile("[\\p{javaLetter}'-]+"); 
	
	/**
	 * For tokens such as "abc.,- def", which will be re-tokenized to "abc" "def"
	 */
	static List<String> tokenizeMultiwordToken(String token) {
		List<String> res = new ArrayList<String>();
		Matcher m = pattern_letters.matcher(token);
		while (m.find()) {
			res.add(m.group());
		}
		return res;
	}
}

File: Alkitab/src/main/java/yuku/alkitab/base/util/QueryTokenizer.kt
package yuku.alkitab.base.util

import java.util.Locale
import java.util.regex.Matcher
import java.util.regex.Pattern

object QueryTokenizer {
    private val oneToken = Pattern.compile("(\\+?)((?:[\"\\u201c\\u201d].*?[\"\\u201c\\u201d]|\\S)+)")
    private val pattern_letters = Pattern.compile("[\\p{javaLetterOrDigit}'-]+")

    /**
     * Convert a query string into tokens. Takes care of the quotes.
     * A single word without quotes means: look for words that contain it. So 'word' matches 'sword'.
     * A single word between quotes means: look for the exact word (aka whole-word). So '"word"' only matches 'word', not 'words' or 'sword'
     * Multiple words between quotes: look for the exact ordering of those words, and all of them needs to be exact. So '"mama papa"' matches 'mama papa' but not 'papa mama' or 'mamas papas' or 'mama papaya'
     *
     * @return List of tokens, starting with the character '+' if it is to be matched in a whole-word/whole-phrase manner.
     * No tokens will be an empty string or "+" (just a plus sign). After the optional '+', there will not be another '+'.
     */
    @JvmStatic
    fun tokenize(query: String?): Array<String> {
        if (query.isNullOrEmpty()) return emptyArray()

        val rawTokens = mutableListOf<String>()

        val matcher = oneToken.matcher(query.lowercase(Locale.getDefault()))
        while (matcher.find()) {
            rawTokens.add((matcher.group(1) ?: "") + matcher.group(2)!!)
        }

        val processed = mutableListOf<String>()
        for (rawTokenOrig in rawTokens) {
            var rawToken = rawTokenOrig
            var plussed = false

            while (true) {
                when {
                    rawToken.isNotEmpty() && rawToken[0] == '+' -> {
                        plussed = true
                        rawToken = rawToken.substring(1)
                    }
                    rawToken.length >= 2 && isQuoteChar(rawToken[0]) && isQuoteChar(rawToken[rawToken.length - 1]) -> {
                        plussed = true
                        rawToken = rawToken.substring(1, rawToken.length - 1)
                    }
                    rawToken.length >= 2 && isQuoteChar(rawToken[0]) -> {
                        plussed = true
                        rawToken = rawToken.substring(1)
                    }
                    else -> break
                }
            }

            if (rawToken.isNotEmpty()) {
                processed.add(if (plussed) "+$rawToken" else rawToken)
            }
        }

        return processed.toTypedArray()
    }

    private fun isQuoteChar(c: Char) = c == '"' || c == '“' || c == '”'

    @JvmStatic
    fun isPlussedToken(token: String) = token.isNotEmpty() && token[0] == '+'

    /**
     * Removes a single '+' from the [token] if exists.
     *
     * @param token may start or not start with '+'
     */
    @JvmStatic
    fun tokenWithoutPlus(token: String): String {
        return if (token.isNotEmpty() && token[0] == '+') token.substring(1) else token
    }

    @JvmStatic
    fun matcherizeTokens(tokens: Array<String>): Array<Matcher> {
        return Array(tokens.size) { i ->
            val token = tokens[i]
            if (isPlussedToken(token)) {
                Pattern.compile("\\b" + Pattern.quote(tokenWithoutPlus(token)) + "\\b", Pattern.CASE_INSENSITIVE).matcher("")
            } else {
                Pattern.compile(Pattern.quote(token), Pattern.CASE_INSENSITIVE).matcher("")
            }
        }
    }

    /**
     * Splits a plussed multiword token (from quoted-phrase input) into its constituent words
     * using the [pattern_letters] pattern. Apostrophe `'` and hyphen `-` are treated as word
     * characters so embedded ones (e.g. `don't`, `self-aware`) stay intact; a standalone `-`
     * surrounded by non-word characters becomes its own "word".
     *
     * @return null if the input produces fewer than two words (i.e. it is not actually a multiword).
     */
    internal fun tokenizeMultiwordToken(token: String): Array<String>? {
        val res = mutableListOf<String>()
        val m = pattern_letters.matcher(token)
        while (m.find()) {
            res.add(m.group())
        }

        if (res.size <= 1) return null
        return res.toTypedArray()
    }
}
